src/lib/base.jl

```diff
@@ -47,6 +47,9 @@ end
   end
 end
 
+@nograd haskey
+@nograd Base.RefValue
```

Comment:
Why wouldn't this need a gradient?

Reply:
Perhaps `@nograd Base.RefValue{Val{f}} where f` be more suited?

I really just want to say that if (after assignment) we are pointing to memory expecting some value, it must keep pointing to that value.

Reply:
` @nograd Base.RefValue{Val{f}} where f` doesn't look like it will do much for that

Reply:
> if (after assignment) we are pointing to memory expecting some value, it must keep pointing to that value.

I'm not sure what you mean by this, can you give more context?

The reason it's an issue is that'd it drop gradients for *all* refs, when things like `RefValue(1)[]` should just propagate gradients as usual. The `Val` version is better since `Val`s can't have gradients anyway, but it'd still be useful to understand why we need it.

Reply:
Fair enough. The motivation comes from trying to run the [vision/mlp/conv.jl](https://github.com/FluxML/model-zoo/blob/master/vision/mnist/conv.jl) example using Zygote and on the GPU and replacing the loss to use MSE instead of crossentropy. The square-ing `2` (I assume) complained about needing an adjoint for constructor `RefValue{Val{2}}(::Float32)`.

Really I'm only interested in making sure we don't take gradients for `Val`s here

Reply:
Ah, I see. That's a bit tricky. I suspect this should be handled by `unbroadcast`; it can dispatch on `Ref{Val}` there and either reduce or throw away gradients as needed. In future we could extend that to allow broadcasting things like `Ref{Int}`, which currently wouldn't work as intended for the same reasons.

Reply:
Ah, I was wondering what could have been triggering that. Makes sense.

Reply:
Should be fixed with https://github.com/FluxML/Zygote.jl/pull/384